pkg/retry/strategy.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package retry
+
+import (
+	"fmt"
+	"time"
+
+	tcontext "github.com/pingcap/dm/pkg/context"
+)
+
+// retryInterval represents enum of retry wait interval
+type retryInterval uint8
+
+const (
+	// LinearIncrease represents increase time wait retry policy, every retry should wait more time depends on increasing retry times
+	LinearIncrease retryInterval = iota + 1
+	// Stable represents fixed time wait retry policy, every retry should wait a fixed time
+	Stable
+)
+
+// DefaultRetryParams define parameters for DefaultRetryStrategy
+// it's a parameters union set of all implements which implement DefaultRetryStrategy
+type DefaultRetryParams struct {
+	RetryCount         int
+	FirstRetryDuration time.Duration
+
+	RetryInterval retryInterval
+
+	// IsRetryableFn tells whether we should retry when operateFn failed
+	// params: (number of retry, error of operation)
+	// return: (bool)
+	//   1. true: means operateFn can be retried
+	//   2. false: means operateFn cannot retry after receive this error
+	IsRetryableFn func(int, error) bool
+}
+
+// Strategy define different kind of retry strategy
+type Strategy interface {
+
+	// DefaultRetryStrategy define retry strategy
+	// params: (retry parameters for this strategy, a normal operation)
+	// return: (result of operation, number of retry, error of operation)
+	DefaultRetryStrategy(ctx *tcontext.Context,
+		params DefaultRetryParams,
+		// TODO: remove `number of retry` in operateFn, we need it now because loader.ExecuteSQL operation need it
+		// operateFn:
+		//   params: (context, number of retry)
+		//   return: (result of operation, error of operation)
+		operateFn func(*tcontext.Context, int) (interface{}, error),
+	) (interface{}, int, error)
+}
+
+// FiniteRetryStrategy will retry `RetryCount` times when failed to operate DB.
+type FiniteRetryStrategy struct {
+}
+
+// DefaultRetryStrategy wait `FirstRetryDuration` before it starts first retry, and then rest of retries wait time depends on RetryInterval.
+// ErrInvalidConn is a special error, need a public retry strategy, so return it to up layer continue retry.
+func (*FiniteRetryStrategy) DefaultRetryStrategy(ctx *tcontext.Context, params DefaultRetryParams,
+	operateFn func(*tcontext.Context, int) (interface{}, error)) (interface{}, int, error) {
+	var err error
+	var ret interface{}
+	i := 0
+	for ; i < params.RetryCount; i++ {
+		ret, err = operateFn(ctx, i)
+		if err != nil {
+			if IsInvalidConnError(err) {
```

Comment:
can we only let `params.IsRetryableFn` to do the check?

@amyangfei what do you think?

Reply:
seems `params.IsRetryableFn` already filters the `ErrInvalidConn` and this check is duplicated

Reply:
some loader execute/query operations don't filter `ErrInvalidConn` for now

Reply:
what about now? @3pointer

Reply:
ErrInvalidConn is used as blacklist here, our whitelist retryable error doesn’t contain this error, so it is automatically filtered?

Reply:
but blacklist does have higher priority here and bumps the specific error here

Reply:
it seems we can remove this logic, or rename `IsInvalidConnError`  to `NotRetryableError`,so we can add blacklist error in this function, may be more clearly?

Reply:
I prefer to remove it for the following reasons
1. The blacklist logic is covered by whitelist logic, it’s duplicated from the logic perspective
2. the invalid conn error will be recorded in the upper level code, the addition warning log here doesn’t carry more information

Reply:
ok, I'll remove it, if we need blacklist in the future, we can add it back